evap/results/tools.py

```diff
@@ -171,42 +178,92 @@ def _calculate_results_impl(course):
+        grouped_results[contributor].extend(results)
+
+    # early return if no rating results exist for this course
+    if not any(result.average for result in grouped_results[contributor] for contributor in grouped_results.keys() if result.question.is_rating_question):
+        return None, None
+
+    contributor_averages = []
+    contributor_distributions = []
+    course_grade_questions_average = None
+    course_grade_questions_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+    course_non_grade_questions_average = None
+    course_non_grade_questions_distribution = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+
+    for contributor, results in grouped_results.items():
+        grade_question_results = [result for result in results if result.question.is_grade_question]
+        grade_questions_average = avg([result.average for result in grade_question_results])
+        grade_questions_distribution = avg_distribution([result.counts for result in grade_question_results])
+        non_grade_question_results = [result for result in results if result.question.is_non_grade_rating_question]
+        non_grade_questions_average = avg([result.average for result in non_grade_question_results])
+        non_grade_questions_distribution = avg_distribution([result.counts for result in non_grade_question_results])
+        if contributor:
+            contributor_averages.append(mix(grade_questions_average, non_grade_questions_average, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+            contributor_distributions.append(avg_distribution(
+                [grade_questions_distribution, non_grade_questions_distribution],
+                [settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE, 1 - settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE]
+            ))
+        else:
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = grade_questions_distribution
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = non_grade_questions_distribution
+
+    contributors_average_grade = avg(contributor_averages)
+    contributors_average_distribution = avg_distribution(contributor_distributions)
+
+    course_non_grade_questions_percentage = 1 - settings.CONTRIBUTIONS_PERCENTAGE - settings.COURSE_GRADE_QUESTIONS_PERCENTAGE
+    average_grade = avg(
+        [contributors_average_grade, course_grade_questions_average, course_non_grade_questions_average],
+        [settings.CONTRIBUTIONS_PERCENTAGE, settings.COURSE_GRADE_QUESTIONS_PERCENTAGE, course_non_grade_questions_percentage]
+    )
+
+    average_distribution = avg_distribution(
+        [contributors_average_distribution, course_grade_questions_distribution, course_non_grade_questions_distribution],
+        [settings.CONTRIBUTIONS_PERCENTAGE, settings.COURSE_GRADE_QUESTIONS_PERCENTAGE, course_non_grade_questions_percentage]
+    )
```

Comment:
now this breaks because avg_distribution is not checking whether there's actually anything inside the distribution.

Reply:
where does it break?

Reply:
every parameter is a distribution (might be all zeros though)

Reply:
yeah it does not break, sorry.